[improve][pip] PIP-445: Add Builder Methods to Create Message-based TableView

namest504 opened the pull request:
This PR introduces **PIP-445: Add Builder Methods to Create Message-based TableView**.

The motivation is to allow users to access the full `Message<T>` object, including metadata (e.g., properties, event time), which is not possible with the current `TableView` API that only returns the deserialized value.

This proposal adds new generic methods to the `TableViewBuilder<T>`:
* `<V> TableView<V> createMapped(Function<Message<T>, V> mapper)`
* `<V> CompletableFuture<TableView<V>> createMappedAsync(Function<Message<T>, V> mapper)`

This allows users to flexibly create a view by providing a custom mapping function. To get a `TableView` of the full `Message<T>` object, a user can pass `Function.identity()`.

---
**Implementation PR:** #24809

mailing list discussion thread: https://lists.apache.org/thread/12jxo0n3njjpm17w2zdo2qvx1bs3y2dk
---

- [x] `doc` <!-- Your PR contains doc changes -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [ ] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

lhotari reviewed (dismissed):
A different approach should be taken. Please check the comment https://github.com/apache/pulsar/pull/24809#pullrequestreview-3334058939 .

The current getRawMessage solution would be a breaking change for existing TableView API users since it would unnecessarily consume more memory and CPU. That's why a different approach must be taken. I've suggested to resolve the issue by adding new methods to the API for returning a TableView with Message<T> as the type parameter.

```java
TableView<Message<T>> createForMessages() throws PulsarClientException;
CompletableFuture<TableView<Message<T>>> createForMessagesAsync() throws PulsarClientException;
```

Would that approach solve your use case?

namest504 committed:
[improve][pip] PIP-445: Add Builder Methods to Create Message-based TableView

lhotari commented:
When I was working on the experiment, https://github.com/lhotari/pulsar/commit/c5a4991, one idea that came into mind is that it would be possible to allow creating a view based on a Function that maps the message to some other object. This would allow using a custom object that takes whatever properties are needed from the message properties and there wouldn't be a need to keep a reference to the complete message instance.
If someone would like to use a message based TableView, passing `java.util.function.Function.identity()` as the function would be a way to use the Message as the value.

namest504 commented:
@lhotari 

Thank you for the continued guidance. I want to confirm my understanding and share my thoughts before proceeding with the implementation.

My understanding is that the goal here is to provide maximum flexibility—evolving beyond a binary choice of payload vs raw message and instead empowering users to transform a Message<T> into any custom object V that fits their specific needs. Is this correct?

I agree this is a very appropriate and powerful extension for this feature.

For example
```java
public class OriginalData {
    String a;
    double b;
}

public class Foo {
    String a;
    double b;
    long eventMetaA;   // from message.getEventTime()
    String eventMetaB; // from message.getProperty()
}

TableView<Foo> view = builder
    .createMapped(message -> {
        OriginalData original = message.getValue();
        return new Foo(
            original.a,
            original.b,
            message.getEventMetaA(),
            message.getEventMetaB()
        );
    });
```
---
I do have two main concerns regarding potential edge cases, and I would appreciate your guidance on them.

1. If a mapper function returns null: What would be the most appropriate way for the TableView to handle this?
2. If the mapper throws an exception: What is the desired behavior in this scenario?

Are these concerns that I should address in the implementation, or are they handled by other parts of the system and not something I need to consider?

---
If my understanding of the direction is correct, I'd like to update the PIP document and the implementation PR accordingly. Please let me know if this is the right way to proceed.

lhotari commented:
> I do have two main concerns regarding potential edge cases, and I would appreciate your guidance on them.
> 
> 1. If a mapper function returns null: What would be the most appropriate way for the TableView to handle this?

In that case, the key is removed from the map since null is considered to be a tombstone value.

> 2. If the mapper throws an exception: What is the desired behavior in this scenario?

Good question. I don't think that this PIP impacts this. If the mapper throws an exception, the same message will get retried, so the table view will get stuck. The table view doesn't have any way to handle "poison pill" messages which fail every time.
I think addressing this is out of context for this PIP.

> Are these concerns that I should address in the implementation, or are they handled by other parts of the system and not something I need to consider?
> 
> If my understanding of the direction is correct, I'd like to update the PIP document and the implementation PR accordingly. Please let me know if this is the right way to proceed.

Makes sense, please go ahead.

namest504 committed:
Update pip-445.md

namest504 committed:
Update pip-445.md

namest504 committed:
update pip-445.mc

namest504 committed:
update pip-445.md

namest504 commented on pip/pip-445.md:
```diff
@@ -0,0 +1,137 @@
+  * A new class will be created that extends `AbstractTableView<Message<T>>` and implements the `TableView<Message<T>>` interface. It will be responsible for storing the full `Message<T>` object for each key, which its `get(key)` method will return.
+
+* **Class `org.apache.pulsar.client.impl.TableViewBuilderImpl<T>`**:
+  * The new `createForMessages` methods will be implemented to instantiate a new `MessageTableViewImpl`.
```
I have updated the PIP to reflect the current design. 🥹

* **Refactoring of `org.apache.pulsar.client.impl.TableViewImpl<T>`**:
  * The existing `TableViewImpl<T>` class is refactored into a new inheritance structure.
  * The core logic is moved to a new abstract base class, `org.apache.pulsar.client.impl.AbstractTableView<T, V>`.
  * To preserve the original functionality, a new simple class `org.apache.pulsar.client.impl.TableView<T>` is introduced, which extends `AbstractTableView<T, T>`.

* **New `MessageMapperTableView` implementation**:
  * `org.apache.pulsar.client.impl.MessageMapperTableView<T, V>`: A new class that extends `AbstractTableView<T, V>`. It implements the logic for the new `createMapped` methods, using a supplied `Function<Message<T>, V>` to transform messages into the values stored in the `TableView`.

* **Class `org.apache.pulsar.client.impl.TableViewBuilderImpl<T>`**:
  * The builder's implementation will be updated to use the new classes.
  * `create()` and `createAsync()` will now instantiate the new `org.apache.pulsar.client.impl.TableView<T>` to provide the classic `TableView` behavior.
  * The new `createMapped()` and `createMappedAsync()` methods will instantiate `MessageMapperTableView<T, V>` with the user-provided mapper function.

lhotari reviewed (commented):
LGTM, please go ahead with the voting for PIP-445

lhotari commented:
> This PR introduces PIP-445, which proposes adding a `getRawMessage()` method to the TableView API.
> 
> This allows users to access the full message metadata (e.g., properties, event time), not just the payload.

Please update the summary in the description of the PR

namest504 commented:
> > This PR introduces PIP-445, which proposes adding a `getRawMessage()` method to the TableView API.
> > This allows users to access the full message metadata (e.g., properties, event time), not just the payload.
> 
> Please update the summary in the description of the PR

Sure, i'll change it now. Thanks your checking! 😊

lhotari committed:
Update mailing list links in pip-445.md

lhotari reviewed (approved):
LGTM

namest504 committed:
Align implementation details with review feedback

- Use the TableViewImpl / AbstractTableViewImpl / MessageMapperTableViewImpl class names
- Clarify message pooling behavior for mapped table views
- Fix leftover createForMessages method name in the downgrade section
